### PR TITLE
Refactor with dedicated registration functions

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #2587 

## Relevant technical choices

* Reworked `withData`'s wrapping component a bit to apply the default date range from the current registry
    * Should be a bit more efficient now with less memory leaks
* See inline comments below  

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/components/higherorder/withData.js [hunk 1]
```diff
@@ -141,6 +147,23 @@ const withData = (
...
+			castArray( dataRequest.context ).forEach( ( context ) => {
+				acc[ context ] = acc[ context ] || [];
+				acc[ context ].push( dataRequest );
+			} );
+			return acc;
+		},
+		{}
+	);
 	// ...and returns another component...
 	class NewComponent extends Component {
```
Comment (aaemnnosttv) [inline assets/js/components/higherorder/withData.js:168]: The changes here will likely be easier to review with whitespace changes hidden
image

File: assets/js/components/higherorder/withData.js [hunk 2]
```diff
@@ -153,100 +176,141 @@ const withData = (
 				moduleRequiringSetup: '',
 			};
 
+			this.handleModuleDataReset = this.handleModuleDataReset.bind( this );
+			this.handleReturnedData = this.handleReturnedData.bind( this );
+			this.addDataRequests = this.addDataRequests.bind( this );
+			this.removeDataRequests = this.removeDataRequests.bind( this );
```
Comment (aaemnnosttv) [inline assets/js/components/higherorder/withData.js:182]: The main reason for these changes is that initially I had simply updated the filter callback to apply the global date range as a default. This worked for the initial render, but because the callbacks were hooked in once from the component's constructor, they didn't respond to date range changes even when the prop was being passed through correctly.

The component also does a better job of cleaning up after itself now; where it previously would have caused potential memory leaks due to its callbacks never being removed.

File: assets/js/components/higherorder/withData.js [hunk 3]
```diff
@@ -153,100 +176,141 @@ const withData = (
...
+
+		addDataRequests() {
+			const { dateRange } = this.props;
+
+			Object.entries( dataRequestsByContext ).forEach(
+				( [ context, dataRequests ] ) => {
+					addFilter(
+						`googlesitekit.module${ context }DataRequest`,
+						`googlesitekit.withData.${ hashRequests( dataRequests ) }`,
+						( contextRequests ) => {
```
Comment (aaemnnosttv) [inline assets/js/components/higherorder/withData.js:226]: This is also a bit more efficient than before. Previously we were adding a filter callback for each request, for each of its contexts. Now we only add a filter once for each _context_ and add all relevant requests at once, for each wrapped component.

File: assets/js/components/higherorder/withData.js [hunk 4]
```diff
@@ -153,100 +176,141 @@ const withData = (
...
+								const { type, identifier } = dataRequest || {};
+								// If a dataRequest's module requires setup, set it in the state.
+								// This will cause the setup incomplete component to be rendered in all cases.
+								if ( TYPE_MODULES === type && ! getModulesData()[ identifier ]?.setupComplete ) {
+									this.setState( { moduleRequiringSetup: identifier } );
+									continue;
+								}
+
+								// Apply default date range if not set.
+								const request = requestWithDateRange( dataRequest, dateRange );
```
Comment (aaemnnosttv) [inline assets/js/components/higherorder/withData.js:238]: Here's where the default date range is applied now, where before it was done in `dataAPI.combinedGet`.

File: assets/js/modules/adsense/index.js [hunk 5]
```diff
@@ -47,17 +41,24 @@ import {
...
-	// IMPORTANT: When updating arguments here, also update the same call in
-	// `provideModuleRegistrations`.
-	Modules.registerModule(
+let isAdBlockerActive = () => {};
+
+export const registerStore = ( registry ) => {
+	registerDataStore( registry );
+	// TODO: fix hack
+	isAdBlockerActive = () => registry.select( STORE_NAME ).isAdBlockerActive();
+};
```
Comment (aaemnnosttv) [inline assets/js/modules/adsense/index.js:58]: I think this is okay as a _temporary workaround_, but it is rather ugly; it's the only thing left that isn't a clean change though. `isAdBlockerActive` here is used within the `checkRequirements` function in the `registerModule` call below. I think the better approach here would be to add support for something like "registry checks" which would be created like a registry control, using a function that works like `createRegistryControl`. That way it could still use the registry if needed (which I think is a reasonable dependency for such a function), or not if not and still be very WP-data like as well as consistent with the overall theme of dependency injection in this PR.

I found a need for a similar pattern with the refactored `CompatibilityChecks` since one of the checks needed access to the registry as well so I used a similar pattern, but just handled the registry part in the same way just without the added API/function/abstraction.

I'm happy to open a new issue for this as a follow-up related issue if we're in agreement about the way forward here?

File: .storybook/config.js [hunk 6]
```diff
@@ -61,8 +61,8 @@ const resetGlobals = () => {
 		isFirstAdmin: true,
```
Comment (aaemnnosttv) [inline .storybook/config.js:61]: First thing's first – the changes here will likely be easier to review with whitespace changes hidden

image

File: assets/js/modules/adsense/index.js (hunk 5 above)
Comment (tofumatt) [reply assets/js/modules/adsense/index.js:58]: > I think the better approach here would be to add support for something like "registry checks" which would be created like a registry control, using a function that works like createRegistryControl. That way it could still use the registry if needed (which I think is a reasonable dependency for such a function), or not if not and still be very WP-data like as well as consistent with the overall theme of dependency injection in this PR.

That sounds like a good path forward to me, I like it.

**Review by tofumatt — APPROVED** [review]
Looks good to me. I like the idea you mentioned for the registry checks, but overall this looks good to me. 👍🏻

File: assets/js/googlesitekit-data.js [hunk 7]
```diff
@@ -25,9 +25,7 @@ if ( typeof global.googlesitekit === 'undefined' ) {
 	global.googlesitekit = {};
 }
 
-if ( typeof global.googlesitekit.data === 'undefined' ) {
-	global.googlesitekit.data = Data;
-}
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/googlesitekit-data.js:30]: Just seemed like an odd precaution to take here. It shouldn't ever be set by anything other than these entry points.

File: assets/js/modules/adsense/index.js (hunk 5 above)
Comment (felixarntz) [reply assets/js/modules/adsense/index.js:58]: Why not pass the `registry` (i.e. `Data`) as parameter to a registered `checkRequirements`? I think that's a simpler way which is reasonable. Not sure why we would want to introduce a separate utility for that. If the function wants to use the `registry`, it can - if not, it doesn't need to.

File: stories/modal-dialog.stories.js [hunk 8]
```diff
@@ -40,6 +40,7 @@ storiesOf( 'Global', module )
 				title={ __( 'Modal Dialog Title', 'google-site-kit' ) }
 				subtitle={ __( 'Modal Dialog Subtitle', 'google-site-kit' ) }
 				provides={ provides }
+				handleConfirm={ global.console.log.bind( null, 'Dialog::handleConfirm' ) }
```
Comment (felixarntz) [inline stories/modal-dialog.stories.js:43]: Why was this change needed?

File: tests/js/utils.js [hunk 9]
```diff
@@ -369,18 +306,10 @@ export const freezeFetch = ( matcher ) => {
...
-	registry.registerStore( MODULES_ANALYTICS, modulesAnalyticsStore );
-	registry.registerStore( MODULES_PAGESPEED_INSIGHTS, modulesPageSpeedInsightsStore );
-	registry.registerStore( MODULES_SEARCH_CONSOLE, modulesSearchConsoleStore );
-	registry.registerStore( MODULES_TAGMANAGER, modulesTagManagerStore );
-	registry.registerStore( MODULES_OPTIMIZE, modulesOptimizeStore );
+	[
+		...allCoreStores,
+		...allModules,
+	].forEach( ( { registerStore } ) => registerStore?.( registry ) );
 };
```
Comment (felixarntz) [inline tests/js/utils.js:353]: Awesome to see all the simplifications in these functions!

**Review by felixarntz — CHANGES_REQUESTED** [review]
There appears to be a problem with this approach which affects the new widget components.

On the dashboard, using a development build, I get this:
Screen Shot 2021-02-02 at 11 20 07 AM

It works fine using a production build, where the Widget API-powered components are not used.

The code overall looks solid, I left a few specific questions.

File: stories/modal-dialog.stories.js (hunk 8 above)
Comment (aaemnnosttv) [reply stories/modal-dialog.stories.js:43]: `handleConfirm` is a required prop.

File: assets/js/googlesitekit-widgets.js [hunk 10]
```diff
@@ -16,28 +16,22 @@
...
 if ( typeof global.googlesitekit === 'undefined' ) {
 	global.googlesitekit = {};
 }
 
-if ( global.googlesitekit.widgets === undefined ) {
-	global.googlesitekit.widgets = Widgets;
-}
-
-// register default widget areas
-domReady( () => {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/googlesitekit-widgets.js:38]: I didn't test translations yet but they should definitely be loaded. Now that translations are loaded per-script using an inline script that runs before the dependent script, we can be sure that translations are immediately available at execution time. It's essentially no different than relying on something in a `Script_Data` global that is a dependency of the script, except the translation loading is automatic.

https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/includes/Core/Assets/Script.php#L119-L137

File: assets/js/components/data/utils/request-with-date-range.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ * Gets a copy of the given data request object with the data.dateRange populated via filter, if not set.
+ * Respects the current dateRange value, if set.
+ *
+ * @since 1.0.0
+ *
+ * @param {Object} originalRequest Data request object.
+ * @param {string} dateRange       Default date range slug to use if not specified in the request.
+ * @return {Object} New data request object.
+ */
+export const requestWithDateRange = ( originalRequest, dateRange ) => {
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/components/data/utils/request-with-date-range.js:29]: I took a look at how often it's actually being called and with all modules active on the main dashboard, it's only 11 times (and again when the date range changes). The legacy data stuff doesn't get retriggered the same way selectors do so it's not an issue 👍

File: assets/js/modules/adsense/index.js (hunk 5 above)
Comment (aaemnnosttv) [reply assets/js/modules/adsense/index.js:58]: @felixarntz that's certainly one solution I considered. I opened #2749 to discuss and address this 👍

**Comment by aaemnnosttv** [comment]
> It works fine using a production build, where the Widget API-powered components are not used.

Thanks @felixarntz – after taking a closer look, the error there was related to recent change where `args.dateRange` was referenced within a `useSelect` where it is no longer set in the `args` object (refactored with actual dates). This resulted in an error in the select function and caused it to return early which had the effect of returning `undefined` which raised the error when the return value was attempted to be destructured. This no longer needed though with the incoming changes from `develop` as the date range was only needed for use with the legacy `getCurrentDateRangeDayCount` utility that was used for passing `num_of_days` in the service URL. This was refactored in #2634.

That said, everything seems to be working fine now, so this should be ready for another pass 👍

**Review by felixarntz — APPROVED** [review]
Awesome!
